Add default ENV variable option with BACKTRACE to turn off backtrace cleaning

dhh opened the pull request:
When debugging framework code, or issues buried in the framework, you often want to just temporarily turn off the backtrace cleaning. Doing so via an ENV variable is an easy reach.

Commit:
Add default ENV variable option with BACKTRACE to turn off backtrace cleaning when debugging framework code

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/initializers/backtrace_silencers.rb.tt:
```diff
@@ -3,5 +3,6 @@
-# You can also remove all the silencers if you're trying to debug a problem that might stem from framework code.
-# Rails.backtrace_cleaner.remove_silencers!
+# You can also remove all the silencers if you're trying to debug a problem that might stem from framework code
+# by setting BACKTRACE=1 before calling your invocation, like "BACKTRACE=1 ./bin/rails test".
```
`rails test` isn’t the best example because it already has a `-b` option to print the full backtrace 😄

dhh replied:
Should have used the runner as an example. And preferably we’d just have one option that did it all.

matthewd commented:
While I do have a general aversion to libraries directly reading ENV, in this case I think I'd like to consider making this a framework-level feature, rather than something that's included in the generated code -- a bit more along the lines of 84eac5dab8b0fe9ee20b51250e52ad7bfea36553. 

@dhh would you accept a follow-up that did that, if someone's up for making a PR?

dhh commented:
That fix was only for tests. I’ve wanted a full backtrace on things like the runner as well. And isn’t this satisfying the idea that it’s not the framework reading directly from ENV? Since it’s in a user generated file?

Don’t really care that much either way, though! This initializer is the natural place to look for changes to the backtrace cleaning, so if this just discussed how to, for example, do —full-backtrace and this option worked for all things bin/rails, I’d be happy too ✌️

Commit:
Merge branch 'master' into env-option-to-turn-off-backtrace-cleaning

Commit:
Use runner as the example

Although I think we should consider dropping -b in tests and just relying on this more general option for sidestepping the backtrace cleaner.

dhh commented:
@matthewd Did you want to take a swing at something else than this? IMO, the backtrace option we have for tests alone is too specific. We should have a top-level option that addresses the entire environment.

matthewd commented:
Agree we should have a mechanism that works everywhere. I had missed that we already ship this as a commented line here, sorry -- given that, I also agree this is a strict improvement 👌🏻

I still feel a bit conflicted about whether it'd be good to have this condition at the library level (it seems like it could be really convenient, but also seems like it'd be a clear layering violation), so as a solution that side-steps that trade-off and immediately gives something that works in apps, this is a nice fit.

Commit:
Merge branch 'master' into env-option-to-turn-off-backtrace-cleaning
